base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,405 @@
+                ConPrintf(StdOut, L" %ls", MemoryUnit[i]);
+
+                return TRUE;
+            }
+        }
+        
+        MemorySize >>= 10; // MemorySize /= 1024;
+    }
+
+    // out of MemoryUnit range.
+    return FALSE;
+}
+
+BOOL EnumProcessAndPrint(BOOL bVerbose)
+{
+    // Load ntdll.dll in order to use NtQuerySystemInformation
+    HMODULE hNtDLL = LoadLibraryW(L"Ntdll.dll");
+    if (!hNtDLL)
+    {
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        return FALSE;
+    }
+
+    NT_QUERY_SYSTEM_INFORMATION PtrNtQuerySystemInformation =
+        (NT_QUERY_SYSTEM_INFORMATION)GetProcAddress(hNtDLL, "NtQuerySystemInformation");
+
+    if (!PtrNtQuerySystemInformation)
+    {
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        FreeLibrary(hNtDLL);
+        return FALSE;
+    }
+
+    // Call NtQuerySystemInformation for the process information
+    ULONG ProcessInfoBufferLength = 0;
+    ULONG ResultLength = 0;
+    PBYTE ProcessInfoBuffer = 0;
+
+    // Get the buffer size we need
+    NTSTATUS Status = PtrNtQuerySystemInformation(SystemProcessInformation, NULL, 0, &ResultLength);
+
+    for (INT Retry = 0; Retry < NT_SYSTEM_QUERY_MAX_RETRY; Retry++)
+    {
+        // (Re)allocate buffer
+        ProcessInfoBufferLength = ResultLength;
+        ResultLength = 0;
+        if (ProcessInfoBuffer)
+        {
+            PBYTE NewProcessInfoBuffer = HeapReAlloc(GetProcessHeap(), 0, ProcessInfoBuffer, ProcessInfoBufferLength);
+            if (NewProcessInfoBuffer)
+            {
+                ProcessInfoBuffer = NewProcessInfoBuffer;
+            }
+            else
+            {
+                // out of memory ?
+                ConResMsgPrintf(StdOut, 0, IDS_OUT_OF_MEMORY);
+                HeapFree(GetProcessHeap(), 0, ProcessInfoBuffer);
+                FreeLibrary(hNtDLL);
+                return FALSE;
+            }
+        }
+        else
+        {
+            ProcessInfoBuffer = HeapAlloc(GetProcessHeap(), 0, ProcessInfoBufferLength);
+            if (!ProcessInfoBuffer)
+            {
+                // out of memory ?
+                ConResMsgPrintf(StdOut, 0, IDS_OUT_OF_MEMORY);
+                FreeLibrary(hNtDLL);
+                return FALSE;
+            }
+        }
+
+        // Query information
+        Status = PtrNtQuerySystemInformation(
+            SystemProcessInformation, ProcessInfoBuffer, ProcessInfoBufferLength, &ResultLength);
+
+        if (Status != STATUS_INFO_LENGTH_MISMATCH)
+        {
+            break;
+        }
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        // tried NT_SYSTEM_QUERY_MAX_RETRY times, or failed with some other reason
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        HeapFree(GetProcessHeap(), 0, ProcessInfoBuffer);
+        FreeLibrary(hNtDLL);
+        return FALSE;
+    }
+
+    // print header
+    PrintHeader();
+
+    PSYSTEM_PROCESS_INFORMATION pSPI;
+    pSPI = (PSYSTEM_PROCESS_INFORMATION)ProcessInfoBuffer;
```

Comment:
Instead of all these intermediate buffers, you could just have used a single pSPI pointer and get rid of your ProcessInfoBuffer. Then in the allocation loop from above you could have just reallocated pSPI up to its correct size.

Reply:
pSPI will be changed when enumerating process, and we still need a pointer pointing to the original buffer in order to free it finally🤣

Reply:
oh, ok then, fine :)